CONFIGURATION.md

```diff
@@ -0,0 +1,179 @@
+Brackets indicate that a parameter is optional.
+For non-list parameters the value is set to the specified default.
+
+Generic placeholders are defined as follows:
+
+* `<boolean>`: a boolean that can take the values `true` or `false`
+* `<duration>`: a duration matching the regular expression `[0-9]+(ms|[smhdwy])`
+* `<filename>`: a valid path in the current working directory
+* `<string>`: a regular string
+* `<secret>`: a regular string that is a secret, such as a password
+* `<regex>`: a regular expression
+
+The other placeholders are specified separately.
+
+### Module
+```yml
+
+  # The protocol over which the probe will take place (http, tcp, dns, icmp)
+  prober: <prober_string>
+
+  # How long the probe will wait before giving up
+  [ timeout: <duration> ]
+
+  [ http: <http_probe> ]
+
+  [ tcp: <tcp_probe> ]
+
+  [ dns: <dns_probe> ]
+
+  [ icmp: <icmp_probe> ]
+
+```
+
+### HTTP
+```yml
+
+  # Accepted status codes for this probe
+  [ valid_status_codes: [ <string> ... ] | default = 2xx ]
+
+  # The HTTP method the probe will use
+   method: [ <string> ]
+
+  # The HTTP headers set for the probe
+  headers: [ <string>: <string> ... ]
+
+  # Whether or not the probe will follow any redirects
+  no_follow_redirects: [ <boolean> ]
+
+  # Probe fails if SSL is present
+  fail_if_ssl: [ <boolean> ]
+
+  # Probe fails if SSL is not present
+  fail_if_not_ssl: [ <boolean> ]
+
+  # Probe fails if response matches regexp
+  fail_if_matches_regexp: [ <regex> ... ]
+
+  # Probe failes if response does not match regexp
+  fail_if_not_matches_regexp: [ <regex> ... ]
+
+  # Configuration for TLS protocol of HTTP probe
+  tls_config:
+
+      # Disable target certificate validation.
+      insecure_skip_verify: [ <boolean> ]
+
+      # The CA cert to use for the targets.
+      ca_file: [ <filename> ]
+
+      # The client cert file for the targets.
+      cert_file: [ <filename> ]
+
+      # The client key file for the targets.
+      key_file: [ <filename> ]
+
+      # Used to verify the hostname for the targets.
+      server_name: [ <string> ]
+
+  # The preferred IP protocol of the HTTP probe
+  preferred_ip_protocol: [ <string> ]
+
+  # The body of the HTTP request used in probe
+  body: [ <string> ]
+
+  # The HTTP basic authentication credentials for the targets.
+  basic_auth:
+      username: [ <string> ]
+      password: [ <secret> ]
+
+  # The bearer token for the targets.
+  bearer_token: [ <secret> ]
+
+  # The bearer token file for the targets.
+  bearer_token_file: [ <filename> ]
+
+  # HTTP proxy server to use to connect to the targets.
+  proxy_url: [ <string> ]
+
+```
+
+### TCP
+
+```yml
+
+# The preferred IP protocol of the TCP probe
+preferred_ip_protocol: [ <string> ]
+
+# The query sent in the TCP probe and the expected associated response
+query_response: [ [ expect: <string>, send: <string> ] ]
+
+# Whether or not TLS is used
+tls: [ <boolean> ]
+
+# Configuration for TLS protocol of TCP probe
+tls_config:
```

Comment:
Rather than listing this multiple times, I'd have a separate tls_config section like we do on prometheus.io